File: kt/minsk/src/main/kotlin/dev/phytolizer/minsk/analysis/syntax/BlockStatementSyntax.kt
package dev.phytolizer.minsk.analysis.syntax

class BlockStatementSyntax(
    val openBraceToken: SyntaxToken,
    val statements: List<StatementSyntax>,
    val closeBraceToken: SyntaxToken,
) : StatementSyntax() {
    override val kind: SyntaxKind
        get() = SyntaxKind.BlockStatement
    override val children: List<SyntaxNode>
        get() = listOf(listOf(openBraceToken), statements, listOf(closeBraceToken)).flatten()
}

File: cs/Minsk/CodeAnalysis/Syntax/BlockStatementSyntax.cs
using System.Collections.Immutable;

namespace Minsk.CodeAnalysis.Syntax;

public sealed class BlockStatementSyntax : StatementSyntax
{
    public SyntaxToken OpenBraceToken { get; }
    public ImmutableArray<StatementSyntax> Statements { get; }
    public SyntaxToken CloseBraceToken { get; }

    public BlockStatementSyntax(
        SyntaxToken openBraceToken,
        ImmutableArray<StatementSyntax> statements,
        SyntaxToken closeBraceToken
    )
    {
        OpenBraceToken = openBraceToken;
        Statements = statements;
        CloseBraceToken = closeBraceToken;
    }

    public override SyntaxKind Kind => SyntaxKind.BlockStatement;

    public override IEnumerable<SyntaxNode> Children
    {
        get
        {
            yield return OpenBraceToken;
            foreach (var statement in Statements)
            {
                yield return statement;
            }

            yield return CloseBraceToken;
        }
    }
}
